frontend/model/captureLogs.js

```diff
@@ -0,0 +1,247 @@
+import { CAPTURED_LOGS, SET_APP_LOGS_FILTER } from '~/frontend/utils/events.js'
+
+/*
+  - giConsole/[username]/limit - the limit of entries.
+  - giConsole/[username]/markerNth - the nth index in which marker log
+  - giConsole/[username]/count - total logs stored
+  - giConsole/[username]/lastEntry - latest log hash
+  - giConsole/[username]/markers - an array of markers at every nth entry
+*/
+
+// Configuration
+const ENTRIES_LIMIT = 5000
+const ENTRIES_MARKER_NTH = 1000
+
+// these are initialized at captureLogsStart()
+let isCapturing = false
+let isConsoleOveridden = false
+let username = null
+let lastEntry = null
+let entriesCount = null
+let appLogsFilter = []
+
+// LS = Local Storage
+const giLSset = (key, value) => localStorage.setItem(`giConsole/${username}/${key}`, value)
+const giLSget = (key) => localStorage.getItem(`giConsole/${username}/${key}`)
+const giLSremove = (key) => localStorage.removeItem(`giConsole/${username}/${key}`)
+
+const getMarkers = () => JSON.parse(giLSget('markers')) || []
+
+function captureLog (type, ...msg) {
+  const logEntry = `${Date.now()}_${Math.floor(Math.random() * 100000)}` // uuid
+
+  // detect when is an Error and capture it properly
+  // ex: uncaught Vue errors or custom try/catch errors.
+  msg = msg.map((m) => m instanceof Error ? (m.stack || m.message) : m)
+
+  giLSset(logEntry, JSON.stringify({
+    type,
+    msg,
+    prev: lastEntry,
+    timestamp: new Date().toISOString()
+  }))
+
+  lastEntry = logEntry
+  entriesCount++
+
+  giLSset('lastEntry', logEntry)
+  giLSset('count', entriesCount)
+
+  // verify if entry is marker
+  if (entriesCount % ENTRIES_MARKER_NTH === 0) {
+    const markers = getMarkers()
+    // Save entry as a marker to be later deleted when logs are too big.
+    markers.push(lastEntry)
+    giLSset('markers', JSON.stringify(markers))
+  }
+
+  // To avoid infinite loop because we log all selector calls, we run sbp calls
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyLogsSize () {
+  let lastEntriesCount = null
+  let markers = null
+
+  // Delete nth oldest logs recursively. This scenario can happen when the
+  // entries limit is changed. Example: There are 400 logs and the limit is 500.
+  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+  // We do it recursively in chunks of 25 until there's only 75 logs again.
+  while (entriesCount >= ENTRIES_LIMIT) {
+    if (entriesCount === lastEntriesCount) {
+      // There are too many entriesCount, however, for some unknown reason
+      // we weren't able to remove them. Call clearLogs to reset everything.
+      clearLogs()
+      console.error('verifyLogsSize(): unable to delete oldest logs, had to clear them!')
```

Comment:
This line will result in an infinite loop.

Reply:
Why do you say that?

It won't. The loop only happens if entriesCount is bigger than the limit. We clear the logs right before calling `console.error`

Reply:
Oh, good point, can you add a comment just above this line explaining that it's safe to call `console.error` here because `clearLogs()` will set `entriesCount` to 0?

Reply:
Done!